File: backend/app/schemas/settings_translation.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.services.translation import SUPPORTED_TARGET_LANGS


class TranslationSettingsRead(BaseModel):
    provider: str = Field(..., description="翻译 Provider（首版固定 google）")
    target_lang: str = Field(..., description="目标语言 BCP-47")
    supported_target_langs: list[str] = Field(
        default_factory=lambda: list(SUPPORTED_TARGET_LANGS),
        description="可选目标语言列表",
    )


class TranslationSettingsUpdate(BaseModel):
    target_lang: str = Field(..., min_length=2, max_length=16)


class TranslationTestResponse(BaseModel):
    ok: bool
    sample: str | None = None
    message: str | None = None

File: frontend/src/lib/settings-translation.ts
export type TranslationSettingsRead = {
  provider: string
  target_lang: string
  supported_target_langs: string[]
}

export type TranslationSettingsUpdate = {
  target_lang: string
}

export type TranslationTestResponse = {
  ok: boolean
  sample?: string | null
  message?: string | null
}

export const TRANSLATION_TARGET_LANG_LABELS: Record<string, string> = {
  "zh-CN": "简体中文",
  "zh-TW": "繁体中文",
  en: "English",
  ja: "日本語",
  ko: "한국어",
}

async function parseErrorMessage(res: Response): Promise<string> {
  try {
    const data: unknown = await res.json()
    if (data && typeof data === "object" && "detail" in data) {
      const d = (data as { detail: unknown }).detail
      if (typeof d === "string") {
        return d
      }
      return JSON.stringify(d)
    }
  } catch {
    // ignore
  }
  return res.statusText || `HTTP ${res.status}`
}

export async function fetchTranslationSettings(): Promise<TranslationSettingsRead> {
  const res = await fetch("/api/settings/translation")
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<TranslationSettingsRead>
}

export async function putTranslationSettings(
  body: TranslationSettingsUpdate
): Promise<TranslationSettingsRead> {
  const res = await fetch("/api/settings/translation", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<TranslationSettingsRead>
}

export async function postTranslationTest(): Promise<TranslationTestResponse> {
  const res = await fetch("/api/settings/translation/test", { method: "POST" })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<TranslationTestResponse>
}
